go/pkg/cas/client.go

```diff
@@ -0,0 +1,110 @@
+// Package cas implements an efficient client for Content Addressable Storage.
+package cas
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+	"google.golang.org/grpc"
+
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/client"
+	repb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+)
+
+// Client is a client for Content Addressable Storage.
+// Create one using NewClient.
+//
+// Goroutine-safe.
+//
+// All fields are considered immutable, and should not be changed.
+type Client struct {
+	// The full name of the RBE instance.
+	InstanceName string
+
+	// The configuration that the client was created with.
+	ClientConfig
+
+	conn       *grpc.ClientConn
+	byteStream bspb.ByteStreamClient
+	cas        repb.ContentAddressableStorageClient
+}
+
+// ClientConfig is a config for Client.
+// See DefaultClientConfig() for the default values.
+type ClientConfig struct {
+	// The connection to RBE to use.
+	//
+	// WARNING: setting this field overrides DialParams completely - DialParams
+	// will be ignored.
+	Conn *grpc.ClientConn
+
+	// Parameters to create a connection.
+	// Ignored if Conn is not nill.
+	DialParams client.DialParams
+
+	// TODO(nodir): add conifg values here.
+	// TODO(nodir): add per-RPC timeouts.
+	// TODO(nodir): add retries.
+}
+
+// DefaultClientConfig returns the default config.
+//
+// To override a specific value:
+//   cfg := DefaultClientConfig()
+//   ... mutate cfg ...
+//   client, err := NewClientWithConfig(ctx, cfg)
+func DefaultClientConfig() ClientConfig {
+	return ClientConfig{
+		DialParams: client.DialParams{
+			Service: "remotebuildexecution.googleapis.com:443",
+		},
+	}
+}
+
+// Validate returns a non-nil error if the config is invalid.
+func (c *ClientConfig) Validate() error {
+	if c.Conn == nil {
+		// Validate DialParams only if Conn is not specified.
+		if c.DialParams.Service == "" {
+			return errors.Errorf("DialParams.Service is unspecified")
+		}
+	}
+
+	return nil
+}
+
+// NewClient creates a new client with the default configuration.
+func NewClient(ctx context.Context, instanceName string) (*Client, error) {
+	return NewClientWithConfig(ctx, instanceName, DefaultClientConfig())
+}
+
+// NewClientWithConfig creates a new client and accepts a configuration.
+func NewClientWithConfig(ctx context.Context, instanceName string, config ClientConfig) (*Client, error) {
+	if instanceName == "" {
+		return nil, fmt.Errorf("instance name is unspecified")
+	}
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid config")
+	}
+
+	conn := config.Conn
+	if conn == nil {
+		var err error
+		if conn, err = client.Dial(ctx, config.DialParams.Service, config.DialParams); err != nil {
+			return nil, errors.Wrap(err, "failed to dial")
+		}
+	}
+
+	client := &Client{
+		InstanceName: instanceName,
+		ClientConfig: config,
```

Comment:
Reading this, we're exposing `DialParams` even if we set both `conn` and `DialParams`. Even if we don't yell a warning, we likely should unset `DialParams` if it's exported.

Reply:
PTAL. I had to make `DialParam` a pointer in order to be able to unset it.
It has caused additional complexity in `Validate` and `NewClientWithConfig` though.

Is this what you meant?

Reply:
Frankly, I am not sure it is better. Now the code reader might think that `DialParams` and `Conn` are mutually exclusive because it is possible to not have `DialParams`.

Reply:
Aren't `DialParams` and `Conn` mutually exclusive? Only one of them should ever be used. 

I'm not sure there's an option to improve both simplicity and correctness. And I'm not convinced that thorough checking of mixes of options that _do_ affect each other is bad. I think it's generally better to have an explicit error than a louder comment, particularly as developers work on more complicated codebases with dependency chains.

Granted, as you mentioned in the other thread, this set of configurations might be changed/migrated substantially (maybe we'll migrate to a hopefully tiny "ConnectionConfig" struct used by CAS and Exec?), so if you think the current setup is confusing/complex, I'm happy to look at this again once the structure is more stable.

Reply:
yes, `DialParams` and `Conn` are mutually exclusive.
yes, checking mixes of options that do affect each other *is* bad.
And I agree that an explicit error is better than a comment.
I think we all agree on these, the question was mostly about the chosen balance.

Regardless, I've removed `DialParams` and made `conn` a required param. I think it is fine both use cases:
- It is certainly easier to the `client` package to use `cas` package now (modulo the circular dep)
- Not a big deal for other packages. They just create a connection with `client.Dial` and then call `cas.NewClient` with the conneciton. Seems fine.

and it addresses all the contention points in this PR. PTAL.